### PR TITLE
Documentation PR

### PR DESCRIPTION
**Pull request by JFaber151**
added documentation and some refactoring. Covering everything but the testProcess package

### CODE REVIEW

File: Faultify.Analyze/RandomValueGenerator.cs [hunk 1]
```diff
@@ -41,30 +41,7 @@ public static object GenerateValueForField(Type type, object reference)
 
             try
             {
-                if (type == typeof(bool))
```
Comment (cdadashovkhandan) [inline Faultify.Analyze/RandomValueGenerator.cs]: Why was this moved to a new method?

File: Faultify.Core/ProjectAnalyzing/ProjectReader.cs [hunk 2]
```diff
@@ -11,25 +11,36 @@ public class ProjectReader : IProjectReader
...
-            return Task.Run<IProjectInfo>(() =>
+            return Task.Run(() =>
             {
-                AnalyzerManager analyzerManager = new AnalyzerManager();
+                return AnalyzeProject(path, progress);
+            });
+        }
+
+        /// <summary>
+        ///     Analye the project and return the results
```
Comment (cdadashovkhandan) [inline Faultify.Core/ProjectAnalyzing/ProjectReader.cs]: Typo

File: Faultify.Cli/Program.cs [hunk 3]
```diff
@@ -51,6 +51,25 @@ private Program(Settings programSettings)
...
+                Logger.Fatal("Couldn't get the Faultify service from the system");
+                Environment.Exit(-1);
+            }
+
+            Directory.CreateDirectory(program.ProgramSettings.OutputDirectory);
+            await program.Run();
+        }
+
+        /// <summary>
+        ///     Perform the setup to create the faultify program
```
Comment (cdadashovkhandan) [inline Faultify.Cli/Program.cs]: Might wanna capitalize `faultify` for good measure

File: Faultify.Analyze/RandomValueGenerator.cs (hunk 1 above)
Comment (cdadashovkhandan) [reply Faultify.Analyze/RandomValueGenerator.cs]: the new method seems to be pretty much the exact same size, so I think moving this to a new method just makes it confusing

File: Faultify.Core/ProjectAnalyzing/ProjectReader.cs [hunk 4]
```diff
@@ -11,25 +11,36 @@ public class ProjectReader : IProjectReader
     {
         public Task<IProjectInfo> ReadProjectAsync(string path, IProgress<string> progress)
```
Comment (cdadashovkhandan) [inline Faultify.Core/ProjectAnalyzing/ProjectReader.cs:12]: If the parent interface doesn't have this commented, you might wanna add a small comment

**Review by cdadashovkhandan — CHANGES_REQUESTED** [review]
Tiny nitpicks here and there, but otherwise good job! Please tend to them, as this is important documentation.

File: Faultify.Analyze/RandomValueGenerator.cs (hunk 1 above)
Comment (JFaber151) [reply Faultify.Analyze/RandomValueGenerator.cs]: Personally, I like the readability of 'oh, it just makes a new reference for that type. And as for how that is done, you've got the other method. But I can change it back

File: Faultify.Injection/CoverageRegistry.cs [hunk 5]
```diff
@@ -26,13 +26,18 @@ public static void Initialize()
             _mmf = MemoryMappedFile.OpenExisting("CoverageFile", MemoryMappedFileRights.ReadWrite);
         }
 
+        /// <summary>
+        ///     Try to write to the mutation coverage fail if the process exits
```
Comment (cdadashovkhandan) [inline Faultify.Injection/CoverageRegistry.cs]: I don't think "fail if the process exits" is necessary here

File: Faultify.Injection/TestCoverageInjector.cs [hunk 6]
```diff
@@ -71,42 +76,70 @@ public void InjectModuleInit(ModuleDefinition toInjectModule)
...
+            MethodDefinition cctor = GetOrMake(moduleInitAttributes, toInjectModule, method.ReturnType);
+
+            if (!IsCallAlreadyDone(method, cctor))
+            {
+                AddMethod(method, cctor);
+            }
+        }
+
+        /// <summary>
+        ///     We get the method definition, if it's empty, we create a new method definition instead
```
Comment (cdadashovkhandan) [inline Faultify.Injection/TestCoverageInjector.cs]: Don't think a "we" is necessary here in either of the sentences.

Comment (cdadashovkhandan) [reply Faultify.Injection/TestCoverageInjector.cs]: Perhaps you could rephrase it as "Get new or existing method definition"

File: Faultify.Injection/TestCoverageInjector.cs [hunk 7]
```diff
@@ -71,42 +76,70 @@ public void InjectModuleInit(ModuleDefinition toInjectModule)
...
+        /// </summary>
+        /// <param name="moduleInitAttributes"></param>
+        /// <param name="moduleType"></param>
+        /// <param name="method"></param>
+        /// <returns></returns>
+        private MethodDefinition GetOrMake(MethodAttributes moduleInitAttributes, ModuleDefinition toInjectModule, TypeReference returnType)
```
Comment (cdadashovkhandan) [inline Faultify.Injection/TestCoverageInjector.cs:95]: maybe just call it "GetCctor" ?

File: Faultify.Injection/TestCoverageInjector.cs [hunk 8]
```diff
@@ -71,42 +76,70 @@ public void InjectModuleInit(ModuleDefinition toInjectModule)
...
+                cctor = new MethodDefinition(".cctor", moduleInitAttributes, returnType);
                 moduleType.Methods.Add(cctor);
             }
 
-            bool isCallAlreadyDone = cctor.Body.Instructions.Any(instruction =>
+            return cctor;
+        }
+
+        /// <summary>
+        ///     We check if the call is already done
```
Comment (cdadashovkhandan) [inline Faultify.Injection/TestCoverageInjector.cs]: No need for "we"

Comment (cdadashovkhandan) [reply Faultify.Injection/TestCoverageInjector.cs]: In general, when writing these comments, mostly focus on the scenario of someone reading this description as they hover over a method name. Therefore, they should just be a quick concise definition that serves as an annotation for the thing they're describing.

File: Faultify.Injection/TestCoverageInjector.cs [hunk 9]
```diff
@@ -71,42 +76,70 @@ public void InjectModuleInit(ModuleDefinition toInjectModule)
...
 
-            // If the method is not called, we can add it
-            if (!isCallAlreadyDone)
-            {
-                ILProcessor ilProcessor = cctor.Body.GetILProcessor();
-                Instruction retInstruction =
-                    cctor.Body.Instructions.FirstOrDefault(instruction => instruction.OpCode == OpCodes.Ret);
-                Instruction callMethod = ilProcessor.Create(OpCodes.Call, method);
+        /// <summary>
+        ///     We add the method, and add any ret instruction if necessary
```
Comment (cdadashovkhandan) [inline Faultify.Injection/TestCoverageInjector.cs]: See previous comment

File: Faultify.Injection/TestCoverageInjector.cs [hunk 10]
```diff
@@ -137,32 +170,33 @@ public void InjectTargetCoverage(ModuleDefinition module)
...
 
-                    ILProcessor processor = method.Body.GetILProcessor();
-
-                    // Insert instruction that loads the meta data token as parameter for the register method.
-                    Instruction assemblyName = processor.Create(OpCodes.Ldstr, method.Module.Assembly.Name.Name);
-
-                    // Insert instruction that loads the meta data token as parameter for the register method.
-                    Instruction entityHandle = processor.Create(OpCodes.Ldc_I4, method.MetadataToken.ToInt32());
+        /// <summary>
+        ///     Get the opcode instructions, and insert them. Including the assmebly name
```
Comment (cdadashovkhandan) [inline Faultify.Injection/TestCoverageInjector.cs]: Typo at "assmebly"

File: Faultify.Injection/TestCoverageInjector.cs [hunk 11]
```diff
@@ -181,25 +215,32 @@ public void InjectTestCoverage(ModuleDefinition module)
...
-                ILProcessor processor = method.Body.GetILProcessor();
-                // Insert instruction that loads the meta data token as parameter for the register method.
-                Instruction entityHandle =
-                    processor.Create(OpCodes.Ldstr, method.DeclaringType.FullName + "." + method.Name);
+        /// <summary>
+        ///     Get the opcode instructions, and insert them. Including the assmebly name
+        /// </summary>
+        /// <param name="method"></param>
+        /// <param name="registerMethodReference"></param>
+        private void InsertInstructions(MethodDefinition method, MethodReference registerMethodReference)
```
Comment (cdadashovkhandan) [inline Faultify.Injection/TestCoverageInjector.cs:232]: Are you sure this comment matches what it's describing?

File: Faultify.TestRunner/MutationTestProject.cs [hunk 12]
```diff
@@ -145,13 +145,22 @@ private TestProjectInfo GetTestProjectInfo(TestProjectDuplication duplication, I
...
                 ModuleDefinition.ReadModule(duplication.TestProjectFile.FullFilePath())
             );
+            LoadInMemory(duplication, projectInfo);
 
-            // Foreach project reference load it in memory as an 'assembly mutator'.
+            return projectInfo;
+        }
+
+        /// <summary>
+        ///     Foreach project reference load it in memory as an 'assembly mutator'.
```
Comment (cdadashovkhandan) [inline Faultify.TestRunner/MutationTestProject.cs]: might want to change `'assembly mutator'`  to `AssemblyMutator`

**Review by cdadashovkhandan — CHANGES_REQUESTED** [review]
I think it's best to follow a more consistent convention in terms of commenting, such as

`/// Attempt to read X and insert it into Y`

instead of 

`/// We try to read X, if it exists, we insert it into Y`

Other than that, good job! The direct code refactors are pretty smart and neat.

File: Faultify.TestRunner/MutationTestProject.cs (hunk 12 above)
Comment (mbartelsm) [reply Faultify.TestRunner/MutationTestProject.cs]:
```suggestion
        ///     Foreach project reference load it in memory as an <see cref="AssemblyMutator"/>.
```

File: Faultify.Injection/TestCoverageInjector.cs (hunk 10 above)
Comment (mbartelsm) [reply Faultify.Injection/TestCoverageInjector.cs]:
```suggestion
        ///     Get the opcode instructions, and insert them. Including the assembly name
```

File: Faultify.Injection/TestCoverageInjector.cs (hunk 9 above)
Comment (mbartelsm) [reply Faultify.Injection/TestCoverageInjector.cs]:
```suggestion
        ///     Add the method, and add any ret instruction if necessary
```

File: Faultify.Injection/TestCoverageInjector.cs (hunk 8 above)
Comment (mbartelsm) [reply Faultify.Injection/TestCoverageInjector.cs]:
```suggestion
        ///     Check if the call is already done
```

File: Faultify.Injection/TestCoverageInjector.cs (hunk 6 above)
Comment (mbartelsm) [reply Faultify.Injection/TestCoverageInjector.cs]:
```suggestion
        ///     Get the method definition for the constructor. If it's empty, create a new method definition instead
```

File: Faultify.Injection/CoverageRegistry.cs (hunk 5 above)
Comment (mbartelsm) [reply Faultify.Injection/CoverageRegistry.cs]:
```suggestion
        ///     Try to write to the mutation coverage fail
```

File: Faultify.Injection/TestCoverageInjector.cs [hunk 13]
```diff
@@ -181,25 +215,32 @@ public void InjectTestCoverage(ModuleDefinition module)
...
+                    }
                 }
+        }
 
-                ILProcessor processor = method.Body.GetILProcessor();
-                // Insert instruction that loads the meta data token as parameter for the register method.
-                Instruction entityHandle =
-                    processor.Create(OpCodes.Ldstr, method.DeclaringType.FullName + "." + method.Name);
+        /// <summary>
+        ///     Get the opcode instructions, and insert them. Including the assmebly name
```
Comment (mbartelsm) [inline Faultify.Injection/TestCoverageInjector.cs]:
```suggestion
        ///     Insert a callback to the register function
```

File: Faultify.Injection/TestCoverageInjector.cs (hunk 7 above)
Comment (mbartelsm) [reply Faultify.Injection/TestCoverageInjector.cs:95]: Cannot change this from the PR, it's up to @JFaber151 to do it locally. All other suggestions have been handled. Remember that for small changes you can make committable code suggestions directly from github

**Pull request merged**